pkg/storage/db.go

```diff
@@ -473,8 +475,125 @@ func (db *DB) Store(ctx context.Context, cloudAssetChanges domain.CloudAssetChan
 		}
 		return err
 	}
-	return tx.Commit()
+	if err = tx.Commit(); err != nil {
+		return err
+	}
+	ver, err := db.GetSchemaVersion(ctx)
+	if err != nil || ver < DualWriteSchemaVersion {
+		return nil
+	}
+	return db.StoreV2(ctx, cloudAssetChanges)
+}
 
+// StoreV2 Currently public to allow testing separately.Storage interface implementation that records to a database using new schema, to replace Store in the future.
+func (db *DB) StoreV2(ctx context.Context, cloudAssetChanges domain.CloudAssetChanges) error {
+	tx, err := db.sqldb.Begin()
+	if err != nil {
+		return err
+	}
+	if err = db.ensureResourceExists(ctx, cloudAssetChanges, tx); err == nil {
+		arnID := resIDFromARN(cloudAssetChanges.ARN)
+		for _, val := range cloudAssetChanges.Changes {
+			for _, ip := range val.PrivateIPAddresses {
+				if strings.EqualFold(added, val.ChangeType) {
+					err = db.assignPrivateIP(ctx, tx, arnID, ip, cloudAssetChanges.ChangeTime)
+				} else {
+					err = db.releasePrivateIP(ctx, tx, arnID, ip, cloudAssetChanges.ChangeTime)
+				}
+				if err != nil {
+					break
+				}
+			}
+			for ix, ip := range val.PublicIPAddresses {
+				hostname := val.Hostnames[ix]
+				if strings.EqualFold(added, val.ChangeType) {
+					err = db.assignPublicIP(ctx, tx, arnID, ip, hostname, cloudAssetChanges.ChangeTime)
+				} else {
+					err = db.releasePublicIP(ctx, tx, arnID, ip, hostname, cloudAssetChanges.ChangeTime)
+				}
+			}
+			if err != nil {
+				break
+			}
+		}
+	}
+	if err != nil {
+		if rollbackErr := tx.Rollback(); rollbackErr != nil {
+			return errors.Wrap(rollbackErr, err.Error()) // so we don't lose the original error
+		}
+		return err
+	}
+	if err = tx.Commit(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (db *DB) ensureResourceExists(ctx context.Context, cloudAssetChanges domain.CloudAssetChanges, tx *sql.Tx) error {
+	// reading pre-requisite to understand this query - https://www.postgresql.org/docs/current/queries-with.html
+	const createResourceQuery string = `
+with sel as (
+    select val.arn_id,
+           val.region,
+           aws_region.id        as aws_region_id,
```

Comment:
nit: for some reason this spacing feels odd to me, I'm more use to SQL like `aws_region.id as aws_region_id`. This is probably coming from `goimport` though so we probably can't avoid it for now.

Reply:
This is intellij sub-language support with auto-formatting specific to PSQL.
Goimports does not "grok" sublanguages.
It is an interesting topic re: formatting SQL as we can't expect everyone to use Jetbrains tools...

Reply:
I see, it can stay but just an observation we should keep in mind is the formatting in the SQL statements is un-uniform throughout the repo. Something we can improve in the future maybe.